### PR TITLE
Harden SecKeyCopyExternalRepresentation against sensitive keys

### PR DESCRIPTION
**Pull request by vcsjones**
The previous approach relied on examining the error code returned from SecKeyCopyExternalRepresentation when a key needed to be exported with a password. Apple has changed the error code which resulted in breaking the detection of sensitive values.

This change looks for the kSecAttrIsSensitive attribute on a key which according to the Apple documentation, "When set to kCFBooleanTrue, the item can only be exported in an encrypted format". This should be less brittle change checking for the error code.

### CODE REVIEW

**Comment by filipnavara** [comment]
I will check it tomorrow morning. There were some odd cases around the old CSSM keys and I don’t quite remember them all.

Thanks for looking into it!

**Comment by vcsjones** [comment]
Yeah, that's why my initial PR was just "handle the new error code" since I would have way more confidence with that in a backport.

The way I see it we have a couple of options.

1. Close this PR and just handle new codes as they happen. This makes it feel too reactionary to me.
2. "Do both". In this PR I eliminated the error handling code path, but I suppose we could do both. Go down the SecItemExport path if either the attribute is present or we get an error code back
3. We try the SecItemExport path on _any_ error.

**Comment by filipnavara** [comment]
I think there’s also a variation of option 4. We can fallback for everything except for a list of known status codes (ie. inverse of what we do now when comparing the status codes). Not saying it’s necessarily a good option, just mentioning it for completeness.

**Comment by filipnavara** [comment]
I did some sanity check first to make sure that the assumptions are correct:
1) Modern implementation (non-CSSM) always have `kSecAttrIsSensitive = false`, `kSecAttrIsExtractable = true`, and `SecKeyCopyExternalRepresentation` essentially cannot fail.
2) Legacy implementation (CSSM; used in certs and keychains) mirrors `CSSM_KEYATTR_SENSITIVE`, `CSSM_KEYATTR_EXTRACTABLE` flags.
3) `CSSMERR_CSP_INVALID_KEYATTR_MASK` error is used to cover two situations:
     a) The key is not extractable (`kSecAttrIsExtractable == false`)
     b) The key is extractable but sensitive (`kSecAttrIsExtractable == true && kSecAttrIsSensitive == true`) and thus password protection is required.

To summarize, it's safe to skip `SecKeyCopyExternalRepresentation` if `kSecAttrIsSensitive` is set to `true` because it's guaranteed to fail. That said, if we are checking the attributes already maybe it would make more sense to move the check higher up and also produce more meaningful error message for `kSecAttrIsExtractable`.

**Review by filipnavara — APPROVED** [review]

**Review by bartonjs — APPROVED** [review]
